### ISSUE TITLE
Require top-level `build-depends` when it occurs in all branches of a conditional

### ISSUE THREAD

**Issue by andreasabel** [haskell/hackage-server#984]
This is the companion issue of https://github.com/haskell/cabal/issues/7768.

See e.g. https://github.com/jvranish/Lenses/issues/3.
Currently hackage does not allow revision of conditionals, and it does not allow new `build-depends` in a revision.
In the situation (1)
```
build-depends: base
  , A
if cond
  build-depends: B >= 0.1 && < 0.4
else
  build-depends: B >= 0.2 && < 0.5
```
the constraints for `B` cannot be revised.  
Thus, the better practice is to put a top-level `B` with extra constraints in the conditional (2):
```
build-depends: base
  , A
  , B >= 0.1 && < 0.5
if cond
  build-depends: B < 0.4
else
  build-depends: B >= 0.2
```
hackage uploads could be required to add a top-level `build-depends` if it appears in all branches.  So (1) would be rejected, and a suggestion of adding `B` to the top-level dependencies could be given:
```
build-depends: base
  , A
  , B
if cond
  build-depends: B >= 0.1 && < 0.4
else
  build-depends: B >= 0.2 && < 0.5
``` 
Enforcing this discipline would help with revisions on hackage that tighten constraints.

**Comment by andreasabel** [comment]
I think I was mistaken, content of conditionals can be revised.  However, conditionals cannot be added or deleted.

The OP looks like this:
```
base (>=4 && <4.15)
    mtl (>=1.1)
    if impl(ghc)
        if impl(ghc>=8.0)
            template-haskell (>=2.11.0.0)
        else
            if impl(ghc>=7.10)
                template-haskell (>=2.10.0.0)
            else
                if impl(ghc>=6.12)
                    template-haskell (>=2.4)
                else
                    template-haskell (>=2.2 && <2.4)
```
> Build with GHC 9.0 fails because of a change in template-haskell (see below).
> A bound template-haskell < 2.17 would be necessary for the versions already released on hackage. 

Such a bound cannot be added to an _existing_ conditional in this case.  We need a _new_ conditional:
```
  if impl (ghc < 9.0)
    template-haskell (< 2.17)
```
If new conditionals could be added if they do not introduce new dependencies (proposal!), then the discipline
> Require top-level build-depends when it occurs in all branches of a conditional

would help revisability.

**Comment by andreasabel** [comment]
A simpler example: To allow `bytestring-0.11` just for `GHC >= 8`, I wanted to add to existing `bytestring` bounds the conditional bound
```
if !impl(ghc >=8.0) { build-depends: bytestring < 0.11 } 
```
as suggested in https://hackage.haskell.org/package/bytestring-0.11.0.0/changelog.
This was rejected.  I had to upload a new version: https://hackage.haskell.org/package/xor-0.0.1.1.

Reconstruction of the case (not all faithful---no sensible revision):
![Screenshot 2022-02-16 at 20 10 57](https://user-images.githubusercontent.com/1155218/154338336-dff308b7-a72b-4f53-836b-567cb853999d.png)

**Comment by gbaz** [comment]
Good example. I'd be fine allowing this.

**Comment by phadej** [comment]
FWIW, adding of conditionals with `build-depends` which don't add new dependencies has already have been discussed, I don't remember whether my and Herbert discussions are recorded though. Maybe in `hackage-cli` repository.

The problem is that someone have to implement it. (And maybe specify it a bit more formally, to ensure that revisions can be undone).

The naive approach of testing all combinations wouldn't work well, as e.g. https://hackage.haskell.org/package/semigroups-0.20 or https://hackage.haskell.org/package/transformers-compat-0.7.1 have over dozen of conditionals, and these are legit use cases.

**Comment by gbaz** [comment]
I think it would be fine to allow new conditionals to mention any dependencies which occur in any existing branch.

**Comment by phadej** [comment]
I forgot to mention that it would be good that specification is not only revertible but also transitive:

I can think of revision path like:

```cabal
-- 1
build-depends: foo
```

```cabal
-- 2
build-depends: foo

if someConditional
  build-depends: foo
```

```cabal
-- 3
if someConditional
  build-depends: foo
```

That doesn't seem right. My concern is that **revisions shouldn't add any new dependencies in any possible configurations**,  they just relax or tighten the constraints. If above is allowed (i.e. adding conditionals with dependencies which are not already there) then revisions could introduce new dependencies for some people. 

There are flags which disable/enable/select between different things, like `HsOpenSSL` / `tls`. If revisions would allow to add `tls`-dependency to openssl-branch, it would make revision making a bit too powerful for my liking. 

I would like the revision from 2 to 3 (and reverse) to be forbidden. 1 <-> 2 are fine.

**Comment by gbaz** [comment]
Fair enough. A good first cut would then be that new conditionals can be added iff they only place constraints on dependencies already listed at the top level, which seems close to the original intent of this ticket.
